From 54a5253bd34af27440caf0ccdc98e8f1bcb7dab2 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Mon, 16 Dec 2013 19:43:40 +0000
Subject: [PATCH 14/26] MinGW-w64: Early create sysroot/mingw top dir

.. it must exist before building GCC as otherwise
the build will deliberately fail.

Left some WIP stuff in regarding the name of this
dir, as we would like to name them:

CT_MULTILIB=y -> mingw
CT_ARCH_64=y  -> mingw64
else          -> mingw32

.. but GCC doesn't support this at present.
---
 scripts/build/cc/100-gcc.sh     | 12 ++++++++++++
 scripts/build/kernel/windows.sh | 16 ++++++++++++++++
 scripts/build/libc/mingw.sh     | 15 +++++----------
 3 files changed, 33 insertions(+), 10 deletions(-)

diff --git a/scripts/build/cc/100-gcc.sh b/scripts/build/cc/100-gcc.sh
index 26ffc7f..c50c0f6 100644
--- a/scripts/build/cc/100-gcc.sh
+++ b/scripts/build/cc/100-gcc.sh
@@ -393,6 +393,18 @@ do_gcc_core_backend() {
         extra_config+=("--with-lipo=${CT_PREFIX_DIR}/bin/${CT_TARGET}-lipo${exeext}")
     fi
 
+    if [ ! "${CT_TARGET/mingw/}" = "${CT_TARGET}" -o ! "${CT_TARGET/cygwin/}" = "${CT_TARGET}" ]; then
+        # This folder must exist otherwise you get an error:
+        # The directory that should contain system headers does not exist:
+        # ${CT_SYSROOT_DIR}/${CT_TARGET_MINGW_SYSROOT_TOP}/include
+        # Linux will create this somewhere else, and also, may want to use
+        # mingw32 and mingw64 as MinGW-w64 itself does now-a-days so until
+        # I decide on these issues this workaround will suffice.
+        if [ ! -d ${CT_SYSROOT_DIR}/${CT_TARGET_MINGW_SYSROOT_TOP}/include ]; then
+            CT_DoExecLog ALL mkdir -p ${CT_SYSROOT_DIR}/${CT_TARGET_MINGW_SYSROOT_TOP}/include
+        fi
+    fi
+
     CT_DoLog DEBUG "Extra config passed: '${extra_config[*]}'"
 
     # Use --with-local-prefix so older gccs don't look in /usr/local (http://gcc.gnu.org/PR10532)
diff --git a/scripts/build/kernel/windows.sh b/scripts/build/kernel/windows.sh
index 5e34420..24911aa 100644
--- a/scripts/build/kernel/windows.sh
+++ b/scripts/build/kernel/windows.sh
@@ -7,6 +7,22 @@ CT_DoKernelTupleValues() {
     # bet mingw32 (require by gcc and mingw-w64)
     CT_TARGET_KERNEL="mingw32"
     CT_TARGET_SYS=
+
+    # May not be the best place - or name - for this!
+    # There seems to be some amount of voodoo around
+    # the naming of this so it may require tweaking.
+    if [ "${CT_MULTILIB}" = "y" ]; then
+        CT_TARGET_MINGW_SYSROOT_TOP="mingw"
+    elif [ "${CT_ARCH_64}" = "y" ]; then
+        CT_TARGET_MINGW_SYSROOT_TOP="mingw64"
+    else
+        CT_TARGET_MINGW_SYSROOT_TOP="mingw32"
+    fi
+    # .. and finally, hack it back to "mingw" because
+    # as gcc/config/i386/mingw32.h is hardcoded:
+    # #define NATIVE_SYSTEM_HEADER_DIR "/mingw/include"
+    # I must speak with Alexey about this.
+    CT_TARGET_MINGW_SYSROOT_TOP="mingw"
 }
 
 do_kernel_get() {
diff --git a/scripts/build/libc/mingw.sh b/scripts/build/libc/mingw.sh
index 63d1ae9..dda6b24 100644
--- a/scripts/build/libc/mingw.sh
+++ b/scripts/build/libc/mingw.sh
@@ -33,11 +33,11 @@ do_libc_start_files() {
 
     CT_DoLog EXTRA "Configuring Headers"
 
-    CT_DoExecLog CFG        \
+    CT_DoExecLog CFG                             \
     "${CT_SRC_DIR}/mingw-w64-v${CT_WINAPI_VERSION}/mingw-w64-headers/configure" \
-        --build=${CT_BUILD} \
-        --host=${CT_TARGET} \
-        --prefix=/usr       \
+        --build=${CT_BUILD}                      \
+        --host=${CT_TARGET}                      \
+        --prefix=/${CT_TARGET_MINGW_SYSROOT_TOP} \
         "${sdk_opts[@]}"
 
     CT_DoLog EXTRA "Compile Headers"
@@ -48,11 +48,6 @@ do_libc_start_files() {
     
     CT_Popd
 
-    # It seems mingw is strangely set up to look into /mingw instead of
-    # /usr (notably when looking for the headers). This symlink is
-    # here to workaround this, and seems to be here to last... :-/
-    CT_DoExecLog ALL ln -sv "usr" "${CT_SYSROOT_DIR}/mingw"
-
     CT_EndStep
 }
 
@@ -65,7 +60,7 @@ do_libc() {
 
     CT_DoExecLog CFG                                                        \
     "${CT_SRC_DIR}/mingw-w64-v${CT_WINAPI_VERSION}/mingw-w64-crt/configure" \
-        --prefix=/usr                                                       \
+        --prefix=/${CT_TARGET_MINGW_SYSROOT_TOP}                            \
         --build=${CT_BUILD}                                                 \
         --host=${CT_TARGET}                                                 \
 
-- 
1.8.5.2

